system/jlib/jevent.cpp

```diff
@@ -1754,6 +1758,913 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+{
+    return new CTimestampFilterTerm(start, end);
+}
+
+IEventVisitFilterTerm* createThreadIdFilterTerm(const char* threadIds)
+{
+    return parseNumberList(threadIds, convertUnsigned, [](__uint64 threadId) { return createThreadIdFilterTerm(threadId); });
+}
+
+IEventVisitFilterTerm* createPathFilterTerm(const char* path)
+{
+    return new CPathFilterTerm(path);
+}
+
+IEventVisitFilterTerm* createFileIdFilterTerm(__uint64 fileId)
+{
+    return new CUnsignedAttributeFilterTerm(EvAttrFileId, fileId);
+}
+
+IEventVisitFilterTerm* createFileIdFilterTerm(const char* ids)
+{
+    return parseNumberList(ids, convertUnsigned, [](__uint64 fileId) { return createFileIdFilterTerm(fileId); });
+}
+
+IEventVisitFilterTerm* createIndexNodeKindFilterTerm(const char* selector)
+{
+    if (!isEmptyString(selector))
+    {
+        if (streq(selector, "0") || strieq(selector, "b") || strieq(selector, "brannch"))
+            return new CUnsignedAttributeFilterTerm(EvAttrNodeKind, 0);
+        if (streq(selector, "1") || strieq(selector, "l") || strieq(selector, "leaf"))
+            return new CUnsignedAttributeFilterTerm(EvAttrNodeKind, 1);
+    }
+    return nullptr;
+}
+
+IEventVisitFilterTerm* createTextAttributeFilterTerm(EventAttr attribute, const char* value)
+{
+    return new CTextAttributeFilterTerm(attribute, value);
+}
+
+IEventVisitFilterTerm* createUnsignedAttributeFilterTerm(EventAttr attribute, __uint64 value)
+{
+    return new CUnsignedAttributeFilterTerm(attribute, value);
+}
+
+IEventVisitFilterTerm* createUnsignedAttributeFilterTerm(EventAttr attribute, __uint64 start, __uint64 end)
+{
+    return new CUnsignedAttributeFilterTerm(attribute, start, end);
+}
+
+IEventVisitFilterTerm* createBooleanAttributeFilterTerm(EventAttr attribute, bool value)
+{
+    return new CBooleanAttributeFilterTerm(attribute, value);
+}
+
+ILogicalEventVisitFilterTerm* createLogicalAndFilterTerm()
+{
+    return new CLogicalAndFilterTerm();
+}
+
+ILogicalEventVisitFilterTerm* createLogicalOrFilterTerm()
+{
+    return new CLogicalOrFilterTerm();
+}
+
+class CEventFileFilter : public CInterfaceOf<IEventVisitor>
+{
+protected:
+    class CVisitQueue : public CInterfaceOf<IEventVisitor>
+    {
+    protected:
+        struct Visit
+        {
+            virtual Continuation flush(IEventVisitor& target) = 0;
+            virtual ~Visit() = default;
+        };
+
+        struct FileVisit : public Visit
+        {
+            StringAttr filename;
+            uint32_t version;
+
+            FileVisit(const char* _filename, uint32_t _version)
+                : filename(_filename), version(_version)
+            {
+            }
+
+            virtual Continuation flush(IEventVisitor& target) override
+            {
+                return (target.visitFile(filename, version) ? IEventVisitor::visitContinue : IEventVisitor::visitSkipFile);
+            }
+        };
+
+        struct EventVisit : public Visit
```

Comment:
The pending list is more complex than I would expect (e.g. files, and events).  Probably worth revisiting in the light of other changes.

E.g. I would probably start with a vector of (attribute, type, value-union)